### PR TITLE
Creates a Janitor Bluespace Light Replacer

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
The janitor is now getting a second bluespace upgrade! The bluespace light replacer is researchable under Advanced Sanitation Tech, and works similar to the Bluespace RPED. Just point at a broken light and click, and it gets replaced on the fly! 

And because having an upgraded and weaker version of the same thing in the same node would be weird, I'm also moving the base light replacer to be a round-start item.

## Why It's Good For The Game
Lights inside of departments rarely get replaced, due to the hassle of getting people to let you in for such a small thing. This change makes it MUCH easier for janitors to replace lights in restricted areas, as if theres a window you can just shoot the lightbulb back on.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

lightrep

(new sprite)

light_base

light_new

https://user-images.githubusercontent.com/73374039/182002944-263350e6-8df9-42a8-9f40-95f63bc249a3.mp4

</details>

## Changelog

:cl:
add: Added the bluespace light replacer
tweak: Moved the standard light replacer to a round-start item
/:cl:

### CODE REVIEW

**Comment by EvilDragonfiend** [comment]
sprite looks a bit weird, but I like the concept.

**Comment by FriendlyContractor** [comment]
Surely this cannot be used to make entire station unhabitable remotely right ?

**Comment by etherware-novice** [comment]
hm?

**Comment by itsmeow** [comment]
> tweak: Moved the standard light replacer to a round-start every departmental lathe item
This should still only be a part of engineering and janitorial department lathes.

**Comment by etherware-novice** [comment]
but why are light replacers restricted to those two in the first place

**Comment by itsmeow** [comment]
Part of playing the game and interacting with others is asking for access to departments, whether it be the workers there or the AI. Encouraging everyone to play isolated job simulator is not good, if your job requires going to other departments, then you can ask HoP, AI, or someone in that department for access instead.

Now everyone will just print light replacers and replace their own departments, which gives engineers/janitors even less job to do when usually they're already pretty under-worked. The bluespace one is a good addition because it allows janitors to replace lights with less access-wrangling, but letting other crew do it themselves just removes part of the game.

File: code/modules/research/designs/power_designs.dm [hunk 1]
```diff
@@ -65,7 +65,18 @@
 	materials = list(/datum/material/iron = 1500, /datum/material/silver = 150, /datum/material/glass = 3000)
 	build_path = /obj/item/lightreplacer
 	category = list("Power Designs")
-	departmental_flags = DEPARTMENTAL_FLAG_SERVICE | DEPARTMENTAL_FLAG_ENGINEERING
+	departmental_flags = DEPARTMENTAL_FLAG_ALL  // everyone should be able to replace their own lights
+
+/datum/design/light_replacer_blue
```
Comment (itsmeow) [inline code/modules/research/designs/power_designs.dm]: `/datum_design/light_replacer_bluespace` is a better typepath for this. Same for the ID

File: code/modules/research/designs/power_designs.dm [hunk 2]
```diff
@@ -65,7 +65,18 @@
...
+	name = "Bluespace Light Replacer"
+	desc = "A device to automatically replace lights from a distance. Refill with working light bulbs"
+	id = "light_replacer_blue"
+	build_type = PROTOLATHE
+	materials = list(/datum/material/iron = 1500, /datum/material/silver = 150, /datum/material/glass = 3000, /datum/material/bluespace = 300)
```
Comment (itsmeow) [inline code/modules/research/designs/power_designs.dm]: Perhaps increase the glass cost? Say, double?

**Comment by etherware-novice** [comment]
oh

File: code/game/objects/items/devices/lightreplacer.dm [hunk 3]
```diff
@@ -229,8 +231,12 @@
 
 /obj/item/lightreplacer/afterattack(atom/T, mob/U, proximity)
 	. = ..()
+
+	var/beam
```
Comment (itsmeow) [inline code/game/objects/items/devices/lightreplacer.dm]: This var is useless, just check `bluespacemode` in both if statements

File: code/game/objects/items/devices/lightreplacer.dm [hunk 4]
```diff
@@ -229,8 +231,12 @@
...
 	if(!proximity)
-		return
+		if(!bluespacemode)
+			return
+		beam = 1
```
Comment (itsmeow) [inline code/game/objects/items/devices/lightreplacer.dm]: This should be `TRUE` defines anyway, but as said `beam` is useless

File: code/game/objects/items/devices/lightreplacer.dm [hunk 5]
```diff
@@ -253,6 +262,15 @@
 /obj/item/lightreplacer/cyborg/janicart_insert(mob/user, obj/structure/janitorialcart/J)
 	return
 
+/obj/item/lightreplacer/bluespace
+	name = "bluespace light replacer"
+	desc = "A modified light replacer that zaps lights into place. Refill with broken or working light bulbs, or sheets of glass."
+	icon_state = "lightreplacer_blue0"
+	bluespacemode = 1
```
Comment (itsmeow) [inline code/game/objects/items/devices/lightreplacer.dm]: `bluespacemode = TRUE`

File: code/game/objects/items/devices/lightreplacer.dm [hunk 6]
```diff
@@ -65,6 +65,8 @@
 	// when we get this many shards, we get a free bulb.
 	var/shards_required = 4
 
+	var/bluespacemode = 0
```
Comment (itsmeow) [inline code/game/objects/items/devices/lightreplacer.dm]: `bluespacemode = FALSE`

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/lightreplacer.dm (hunk 3 above)
Comment (etherware-novice) [reply code/game/objects/items/devices/lightreplacer.dm]: actually, it makes it so it doesnt play the beam effects if youre standing close to a lightbulb

Comment (etherware-novice) [reply code/game/objects/items/devices/lightreplacer.dm]: wait no it is useless i can check proximity twice

**Comment by etherware-novice** [comment]
oh god damnit indentation error again

**Comment by FriendlyContractor** [comment]
I was hoping it will be obvious for others how this is extremely exploitable and can be very easily used to grief the entire round but nobody pointed it out yet so i better say it out loud least this gets merged.

Lightbulbs can be injected with 5u of plasma to make them rigged which causes them to explode and create plasma fire when powered. Normally this is fine since you need to turn off the light first or you will be the one blown up limiting its uselfullnes **HOWEVER** with bluespace device you can do this remotely through camera console  and cause explosions and plasma fires everywhere there are lightbulbs and cameras.

**Comment by etherware-novice** [comment]
you can inject lights with plasma?!

File: code/game/objects/items/devices/lightreplacer.dm [hunk 7]
```diff
@@ -229,7 +231,9 @@
 
 /obj/item/lightreplacer/afterattack(atom/T, mob/U, proximity)
 	. = ..()
-	if(!proximity)
+
+	var/beam
```
Comment (Sarchutar) [inline code/game/objects/items/devices/lightreplacer.dm]: this var is useless now

Comment (etherware-novice) [reply code/game/objects/items/devices/lightreplacer.dm]: oops ya

**Comment by FriendlyContractor** [comment]
I hope i am wrong but im getting a feeling this is malcious PR trying to intentionally introduce game breaking exploit

**Comment by etherware-novice** [comment]
i didnt know about the plasma lights!

**Comment by Mat05usz** [comment]
> I hope i am wrong but im getting a feeling this is malcious PR trying to intentionally introduce game breaking exploit

Nothing about this or any other their PR looks like malice, in my opinion, why jumping to such conclusions right away?

But yeah, despite the dumbass malice accusation, they have a point, you shouldn't be able to remotely replace lightbulbs with plasma filled ones.

**Comment by etherware-novice** [comment]
ill just make it unemaggable

**Comment by Mat05usz** [comment]
> ill just make it unemaggable

It's not about emag, you can inject plasma into a lightbulb, then place it inside the replacer and use it like that.

**Comment by etherware-novice** [comment]
oh..

**Comment by Mat05usz** [comment]
I'm not sure about the best way to solve this, maybe just make it unable to accept rigged flashbulbs?

**Comment by etherware-novice** [comment]
hey wait it doesnt even use the same light thing i think it just adds to a int of lights

**Comment by Fronsis** [comment]
I don't see how would you use this to create plasma-fires through cams and it's kinda rude to straight up accuse OP for trying to make an exploit???

Sure, if you do it with a regular lightbulb in-hand you can do the plasma trick but i don't think you can inject plasma into the **light replacer** itself meaning that all those bulbs/tubes **can't** be injected pre-placing them since they exist within the light replacer, unless you can inject plasma to the light replacer itself which.. it would be kinda weird but i guess it could be fixed to not being able to do so.

**Comment by etherware-novice** [comment]
> I don't see how would you use this to create plasma-fires through cams and it's kinda rude to straight up accuse OP for trying to make an exploit???
> 
> Sure, if you do it with a regular lightbulb in-hand you can do the plasma trick but i don't think you can inject plasma into the **light replacer** itself meaning that all those bulbs/tubes **can't** be injected pre-placing them since they exist within the light replacer, unless you can inject plasma to the light replacer itself which.. it would be kinda weird but i guess it could be fixed to not being able to do so.

you *can* make plasma filled lights by emagging the light replacer, so i just made it un-emaggable

**Comment by Mat05usz** [comment]
I thought you can fill lights and then put them into the device?

**Comment by etherware-novice** [comment]
does it?? i didnt see anything like that in the code

**Comment by etherware-novice** [comment]
yea, only the bluespace varient is emag proof

**Review by PowerfulBacon — DISMISSED** [review]
the sprite is ugly

**Review by itsmeow — APPROVED** [review]

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**